src/layers/scenegraph-layer/scenegraph-layer.js

```diff
@@ -0,0 +1,312 @@
+import {ScenegraphLayer as DeckScenegraphLayer} from '@deck.gl/mesh-layers';
+import {load} from '@loaders.gl/core';
+import {GLTFScenegraphLoader} from '@luma.gl/addons';
+
+import Layer from '../base-layer';
+import memoize from 'lodash.memoize';
+import ScenegraphLayerIcon from './scenegraph-layer-icon';
+
+export const scenegraphRequiredColumns = ['lat', 'lng'];
+export const scenegraphOptionalColumns = ['altitude'];
+
+export const scenegraphPosAccessor = ({lat, lng, altitude}) => d => [
+  // lng
+  d.data[lng.fieldIdx],
+  // lat
+  d.data[lat.fieldIdx],
+  // altitude
+  altitude && altitude.fieldIdx > -1 ? d.data[altitude.fieldIdx] : 0
+];
+
+export const scenegraphPosResolver = ({lat, lng, altitude}) =>
+  `${lat.fieldIdx}-${lng.fieldIdx}-${altitude ? altitude.fieldIdx : 'z'}`;
+
+export const scenegraphVisConfigs = {
+  radius: 'radius',
+  fixedRadius: 'fixedRadius',
+  opacity: 'opacity',
+  colorRange: 'colorRange',
+  radiusRange: 'radiusRange',
+  //
+  sizeScale: 'sizeScale',
+  angleX: 'angleX',
+  angleY: 'angleY',
+  angleZ: 'angleZ'
+};
+
+export default class ScenegraphLayer extends Layer {
+  constructor(props) {
+    super(props);
+
+    this.registerVisConfig(scenegraphVisConfigs);
+    this.getPosition = memoize(scenegraphPosAccessor, scenegraphPosResolver);
+
+    // prepare layer info modal
+    // this._layerInfoModal = IconInfoModalFactory();
+    // this.getSvgIcons();
+  }
+
+  get type() {
+    return '3D';
+  }
+
+  get requiredLayerColumns() {
+    return scenegraphRequiredColumns;
+  }
+
+  get optionalColumns() {
+    return scenegraphOptionalColumns;
+  }
+
+  get columnPairs() {
+    return this.defaultPointColumnPairs;
+  }
+
+  get layerIcon() {
+    return ScenegraphLayerIcon;
+  }
+
+  get visualChannels() {
+    return {
+      ...super.visualChannels,
+      size: {
+        ...super.visualChannels.size,
+        range: 'radiusRange',
+        property: 'radius',
+        channelScaleType: 'radius'
+      }
+    };
+  }
+
+  get layerInfoModal() {
+    return {
+      id: 'iconInfo',
+      template: this._layerInfoModal,
+      modalProps: {
+        title: 'How to draw icons'
+      }
+    };
+  }
+
+  static findDefaultLayerProps({fieldPairs, fields}) {
+    if (!fieldPairs.length) {
+      return [];
+    }
+
+    const iconFields = fields.filter(({name}) =>
+      name
+        .replace(/[_,.]+/g, ' ')
+        .trim()
+        .split(' ')
+        .some(seg => ({}.icon.some(t => t.includes(seg))))
+    );
+
+    if (!iconFields.length) {
+      return [];
+    }
+
+    // create icon layers for first point pair
+    const ptPair = fieldPairs[0];
+
+    const props = iconFields.map(iconField => ({
+      label: '???',
+      columns: {
+        lat: ptPair.pair.lat,
+        lng: ptPair.pair.lng
+      },
+      isVisible: true
+    }));
+
+    return props;
+  }
+
+  // TODO: fix complexity
+  /* eslint-disable complexity */
+  formatLayerData(_, allData, filteredIndex, oldLayerData, opt = {}) {
```

Comment:
`formatLayerData` only got called when filters changed, or color, radius, size by value is changed. its output should be a new data array or attributed accessors. If `sizeScale `,`angleX`, `angleY`, `angleZ` is directly passed to Deckgl layer, you don't have to calculate them here. Just directly pass them in the `renderLayer`function below